### PR TITLE
KAFKA-7443: OffsetOutOfRangeException in restoring state store from changelog topic when start offset of local checkpoint is smaller than that of changelog topic

### PR DESCRIPTION
**Pull request by linyli001**
See https://issues.apache.org/jira/browse/KAFKA-7443 for details.
The fix set this state partition to "NO_CHECKPOINT" when the offset in local checkpoint file has expired and older than the current start offset of changelog topic, thus making this task to restore local state from the current beginning offset of changelog topic, avoiding falling into the infinite loop caused by this exception.

### CODE REVIEW

**Comment by mjsax** [comment]
Thanks for the PR @linyli001. Can we add a unit (or integration) test for this case?

**Comment by mjsax** [comment]
There is a checkstyle error:
```
ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11@2/streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:112:1: File contains tab characters (this is the first instance). [FileTabCharacter]
```

**Comment by vvcephei** [comment]
The test results are gone already, for some reason...

**Comment by vvcephei** [comment]
Retest this, please.

**Comment by mjsax** [comment]
@linyli001 Can we add a test for this case?

**Review by bbejeck — COMMENTED** [review]
The change looks good to me, but I agree with @mjsax we'll need an integration test for this.

**Comment by linyli001** [comment]
@mjsax Ok, I'll add an integration test. But the scenario only triggers under certain conditions: the changelog topic deleted the expired segments and updated the new starting offset, which is greater than the offset recorded in local checkpoint file. I'm confused how to simulate this scenario in my test case, so could you give me some suggestions?

**Comment by bbejeck** [comment]
@linyli001 I see your point, just off the top of my head without looking at the code, could you simulate the condition better using mocks? In that case, you'd only need a unit test.

**Comment by mjsax** [comment]
I agree that it's a little tricky... A mocked unit test should be possible for sure as suggested by Bill.

For an integration test, you could create a topic with very small segment size, write data into it, and user purgeDataAPI to delete old records. Then, a checkpoint file with old offsets is written and afterwards Streams is started and we check if it recovers correctly.

Does this help? Thoughts?

**Comment by vvcephei** [comment]
I think we'd need to see the unit test to decide if it's complete enough, but it feels like enough to me...

The fundamental question is whether you trust the client to throw that exception when the offset is ... out of range. If that's the specified behavior of the client, then we don't need an integration test, we just need to make sure that Streams properly clears the offset when it gets the exception.

In other words, we should be able to test it satisfactorily with a mock.

**Comment by linyli001** [comment]
Ok. I will add the test case according to @mjsax.  Thanks!

**Comment by linyli001** [comment]
The unit test has some problems, and I'll check and push a new commit latter.

File: clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java [hunk 1]
```diff
@@ -188,6 +191,12 @@ public synchronized void unsubscribe() {
             if (!subscriptions.isPaused(entry.getKey())) {
                 final List<ConsumerRecord<K, V>> recs = entry.getValue();
                 for (final ConsumerRecord<K, V> rec : recs) {
+                    if (beginningOffsets.get(entry.getKey()) != null && beginningOffsets.get(entry.getKey()) > subscriptions.position(entry.getKey())) {
+                        log.info("poll offset {} less than changelog topic start offset {}, throw OffsetOutOfRangeException!", subscriptions.position(entry.getKey()), beginningOffsets.get(entry.getKey()));
+                        RuntimeException exception = new OffsetOutOfRangeException(Collections.singletonMap(entry.getKey(), subscriptions.position(entry.getKey())));
```
Comment (mjsax) [inline clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java]: variable exception unnecessary

File: clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java [hunk 2]
```diff
@@ -188,6 +191,12 @@ public synchronized void unsubscribe() {
             if (!subscriptions.isPaused(entry.getKey())) {
                 final List<ConsumerRecord<K, V>> recs = entry.getValue();
                 for (final ConsumerRecord<K, V> rec : recs) {
+                    if (beginningOffsets.get(entry.getKey()) != null && beginningOffsets.get(entry.getKey()) > subscriptions.position(entry.getKey())) {
+                        log.info("poll offset {} less than changelog topic start offset {}, throw OffsetOutOfRangeException!", subscriptions.position(entry.getKey()), beginningOffsets.get(entry.getKey()));
```
Comment (mjsax) [inline clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java]: Remove `changelog` -- can be any topic

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreChangelogReaderTest.java [hunk 3]
```diff
@@ -157,6 +157,36 @@ public void shouldRecoverFromInvalidOffsetExceptionAndFinishRestore() {
...
+        final int startOffset = 5;
+        assignPartition(messages, topicPartition);
+        consumer.updateBeginningOffsets(Collections.singletonMap(topicPartition, (long) startOffset));
+        consumer.updateEndOffsets(Collections.singletonMap(topicPartition, (long) (messages + startOffset)));
+
+        addRecords(messages, topicPartition, startOffset);
+        consumer.assign(Collections.<TopicPartition>emptyList());
+
+        final StateRestorer stateRestorer = new StateRestorer(topicPartition, restoreListener, 1L, Long.MAX_VALUE, true,
+                "storeName");
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreChangelogReaderTest.java:178]: nit: formatting -- use one line per parameter:
```
final StateRestorer stateRestorer = new StateRestorer(
    topicPartition,
    restoreListener,
    1L,
    Long.MAX_VALUE,
    true,
    "storeName");
```

Maybe introduce variable `long expiredCheckpoint = 1L;` and pass into `StateRestorer` to make this clearer?

**Review by mjsax — COMMENTED** [review]
Thanks for adding the test. Some nits. Overall LGTM.

File: clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java (hunk 1 above)
Comment (linyli001) [reply clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java]: Add the OffsetOutOfRangeException here is to deal with the scenario that the beginning offset for changelog topic is greater than subscription offset. Without this exception, we can't simulate the test case and thus validate the fix has resolved this exception.

Comment (mjsax) [reply clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java]: Adding the exception is fine, but you can just throw it directly:
```
throw new OffsetOutOfRangeException(...)`
```

Not need to assign it to variable first :)

Comment (linyli001) [reply clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java]: So do you mean to use new OffsetOutOfRangeException() instead of Collections.singletonMap(entry.getKey(), subscriptions.position(entry.getKey()))?

Comment (linyli001) [reply clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java]: Ok. I'll modify.

**Review by vvcephei — APPROVED** [review]
This looks good to me, @linyli001 .

Thanks!

**Comment by vvcephei** [comment]
@mjsax , is this ready to merge?

**Comment by mjsax** [comment]
Thanks for the PR @linyli001!

Merged to `trunk` and cherry-picked to `2.1`, `2.0`, and `1.1`.

**Comment by linyli001** [comment]
@mjsax I have some problems when sync with the latest change. I can see my fix has been merged into the trunk branch, but when I folk the apache/kafka to my GitHub, I can't found the latest change. Any steps am I missing?

**Comment by vvcephei** [comment]
Hi @linyli001 ,

IIRC, you have to check out your trunk branch, then pull from apache/kafka's trunk, then push it to your fork's trunk.

Is that what you're already trying to do?

**Comment by linyli001** [comment]
@vvcephei Sorry I don't quite understand how to do it. I have cloned the git from my fork to local, but run "git pull origin trunk" not includes the latest change. I think it just pull from my fork's trunk, but I don't know how to pull from apache/kafka's trunk locally? Can you show me the git command? Thanks!

**Comment by mjsax** [comment]
If you fork the repo, your copy on github won't be updated automatically. You need to pull the latest changes from kafka repo into your local one (and than you can push those back to you copy on github).

Assume you have an alias called `kafka-github` (you can create one with `git add remote...` command) and assume you are on local branch `trunk`:
```
git fetch kafka-github              // get latest changes
git rebase kafka-github trunk       // update local trunk to latest changes
git push origin trunk               // push the changes to you github copy

**Comment by linyli001** [comment]
@mjsax Thanks for the guide! I can update my folk now.
